Present the actual error message from raise_error

genehsu opened the pull request:
Fixes rspec/rspec#137 

Display the actual error message of the raised error when a raise_error
assertion fails.  The failure message will now include the error message
on a separate line followed by the backtrace.  If there is no error
message, only the backtrace will be shown.

If this approach is not correct, then please close this PR without prejudice 😃

genehsu committed:
Present the actual error message from raise_error

Display the actual error message of the raised error when a raise_error
asserton fails.  The failure message will now include the error message
on a separate line followed by the backtrace.  If there is no error
message, only the backtrace will be shown.

JonRowe commented on lib/rspec/matchers/built_in/raise_error.rb:
```diff
@@ -238,10 +238,18 @@ def given_error
+            ", got #{description_of(@actual_error)}",
+            message_text,
+            [" with backtrace:", *backtrace].zip(["\n  # "] * backtrace.size)
+          ].flatten.join('')
```
This array could probably just be interpolation now, as the intent of the original map is taken care of by the zip.

JonRowe commented on spec/rspec/matchers/built_in/raise_error_spec.rb:
```diff
@@ -109,7 +109,31 @@ def to_s
+  end
+
+  it "fails without a message if a different error instance is thrown from the one that is expected" do
+    s = StandardError.new("Error 1")
```
Rather than `s` can you use a proper name like `error`

JonRowe commented on lib/rspec/matchers/built_in/raise_error.rb:
```diff
@@ -238,10 +238,18 @@ def given_error
+            message_text = ''
+          else
+            message_lines = @actual_error.message.split(/\n/)
+            message_text = [" with message:", *message_lines].zip(["\n  # "] * message_lines.size)
```
given that the magic string `"\n  # "` is now used multiple times it might be best to assign it and reuse it

JonRowe reviewed (changes requested):
Thanks for tackling this, can you check out these couple of suggestions?

genehsu commented:
@JonRowe 
> Thanks for tackling this, can you check out these couple of suggestions?

Thanks for the review. I believe I've addressed your suggestions with clearer code. Let me know if you have any other suggestions to the code or output.

JonRowe commented:
Migrated to the monorepo as rspec/rspec#128
